livechatsnew/consumers.py

```diff
@@ -0,0 +1,115 @@
+import json
+
+import pydantic
+from bson import ObjectId
+from datetime import datetime
+from django.core.cache import cache
+from collections import defaultdict
+from channels.db import database_sync_to_async
+from channels.generic.websocket import AsyncWebsocketConsumer
+from livechatsnew.utils import mongo_conversations
+from livechatsnew.schemas import Message
+from pydantic import parse_obj_as
+
+
+class ChatConsumer(AsyncWebsocketConsumer):
+    room_connection_counts = defaultdict(lambda: 0)
+
+    @database_sync_to_async
+    def get_conversation(self):
+        conversations = mongo_conversations()
+        conv = conversations.find_one({"_id": ObjectId(self.room_name)})
+        initiator = conv.get("initiator_id")
+        receiver = conv.get("receiver_id")
+        return conv, initiator, receiver
+
+    @database_sync_to_async
+    def update_conversation(self, messages):
+        """Function for updating live-chat messages data"""
+        conversations = mongo_conversations()
+        conversations.update_one({"_id": ObjectId(self.room_name)}, {"$push": {"messages": messages}})
+
+    async def create_message_data(self, message):
+        """Function for creating messages available for caching and sending"""
+        conversation = await self.get_conversation()
+        sender = self.scope["user"]
+        message = {
+            "msg_text": message["message"],
+            "timestamp": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
+            "sender_id": sender.id,
+            "sender_data": str(sender),
+            "conversation_id": str(conversation[0].get("_id")),
+
+        }
+        return message
+
+    async def cache_data(self, message):
+        """Function for adding messages and joining data to cache for 10 hours"""
+        _message = await self.create_message_data(message)
+        chat_value = cache.get(_message["conversation_id"])
+        if chat_value:
+            chat_value = str(_message) + "," + chat_value
+        else:
+            chat_value = str(_message)
```

Comment:
Concatenating message strings for caching might not be the most efficient or reliable method, especially when it comes to parsing them back into objects

Reply:
Changed way to save messages, now messages ii be saved as list in redis